### DISCUSSION TITLE
ListBox Controlled Keyboard Selection

### DISCUSSION THREAD

**Discussion by BrianHung** [adobe/react-spectrum#6932] [Q&A]
I'm using the listbox component to implement an autocomplete menu, something like Notion's.

The listbox is embedded and should be controlled by the text editor. So whenever arrow keys are pressed down, I want to forward that to the listbox and change the selected keys. Is there a way to do this properly? 

I tried adding a ref to the listbox and re-dispatching the editor event, but it doesn't seem to affect the listbox even though `onKeyDown` within `useSelectableCollection` is triggered. Any thoughts?

**Comment by snowystinger** [comment]
Looks like for Combobox, we pass it along on the ListBox context https://github.com/adobe/react-spectrum/blob/dcf312bc21ac08e9202e110f10572d4df47784b1/packages/react-aria-components/src/ComboBox.tsx#L197

Which is this value in particular https://github.com/adobe/react-spectrum/blob/dcf312bc21ac08e9202e110f10572d4df47784b1/packages/%40react-aria/combobox/src/useComboBox.ts#L369

I'm not sure why we don't expose that in the props on the ListBox component itself, I'll see if anyone else remembers or if it is a bug. In the meantime, you should be able to model yourself after Combobox for this.

**Reply by BrianHung** [reply]
`shouldUseVirtualFocus` does look exposed on `ListBox` but not sure if there's a proper way to dispatch arrow or enter events if its coming from contenteditable.

**Reply by BrianHung** [reply]
Do I have to use a keyboard delegate thing? How does the ComboBox dispatch arrow events?

**Comment by BrianHung** [comment]
Dug into the listbox tests and saw that it requires a keydown _and_ a keyup.

https://github.com/adobe/react-spectrum/blob/3555336a951aed44bb51e55c131dcf7176fc4b59/packages/react-aria-components/test/ListBox.test.js#L57-L58

```
listBoxRef.current!.dispatchEvent(new KeyboardEvent('keydown', event));
listBoxRef.current!.dispatchEvent(new KeyboardEvent('keyup', event));
```

Works!
